crates/astria-conductor/src/data_availability/mod.rs

```diff
@@ -130,337 +228,252 @@ impl Reader {
-    rollup_namespace: Namespace,
-) -> eyre::Result<Vec<SequencerBlockSubset>> {
-    // spawn the verification tasks
-    let mut verification_tasks = verify_all_blobs(sequencer_blobs, &block_verifier);
-
-    let (assembly_tx, assembly_rx) = mpsc::channel(256);
-    let block_assembler = task::spawn(assemble_blocks(assembly_rx));
-
-    let mut fetch_and_verify_rollups = JoinSet::new();
-    while let Some((block_hash, verification_result)) = verification_tasks.join_next().await {
-        match verification_result {
-            Err(e) => {
-                let error = &e as &(dyn std::error::Error + 'static);
-                warn!(
-                    block_hash = %telemetry::display::hex(&block_hash),
-                    error,
-                    "task verifying sequencer data retrieved from celestia failed; dropping block",
-                );
-            }
-            Ok(Err(e)) => {
-                let error: &(dyn std::error::Error + 'static) = e.as_ref();
-                warn!(
-                    block_hash = %telemetry::display::hex(&block_hash),
-                    error,
-                    "task verifying sequencer data retrieved from celestia returned with an \
-                     error; dropping block"
-                );
-            }
-            Ok(Ok(data)) => {
-                fetch_and_verify_rollups.spawn(
-                    fetch_rollup_blob_and_forward_to_assembly(
-                        client.clone(),
+impl Stream for ReconstructedBlockStream {
+    type Item = ReconstructedBlock;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        use futures::{
+            ready,
+            FutureExt as _,
+        };
+        let this = self.project();
+
+        // First up, try to spawn off as many futures as possible by filling up
+        // our queue of futures.
+        while this.in_progress_queue.len() < *this.max {
+            match this.heights.pop_front() {
+                Some(height) => this.in_progress_queue.push(
+                    fetch_blocks_at_celestia_height(
+                        this.client.clone(),
+                        this.verifier.clone(),
                         height,
-                        data,
-                        rollup_namespace,
-                        assembly_tx.clone(),
+                        *this.sequencer_namespace,
+                        *this.rollup_namespace,
                     )
-                    .in_current_span(),
-                );
+                    .boxed(),
+                ),
+                None => break,
             }
         }
-    }
 
-    // ensure that the last sender goes out of scope so that block assembler's exit condition fires
-    std::mem::drop(assembly_tx);
+        // Attempt to pull the next value from the in_progress_queue
+        let res = this.in_progress_queue.poll_next_unpin(cx);
+        if let Some(val) = ready!(res) {
```

Comment:
is what's returned from this queue ordered by block height? it appears no?

Reply:
The stream is unordered. The ordering is done by the block cache. Since sequencer heights need not be in order (i.e. a sequencer height S can appear in a celestia height C+1 while a sequencer height S+1 can appear in celestia height C) enforcing an ordering in the futures does not make sense IMO